### PR TITLE
Added a feature for native projects and fixed a bug for tar distributions

### PR DESCRIPTION
**Pull request by blu3r4y**
Those two commits are independent of each other.

One enables to pass the target argument to gnu make and the other fixes an issue where tar distributions were not removed by mx clean.

### CODE REVIEW

**Comment by blu3r4y** [comment]
I restructured the code, such that a method within `NativeTARDistribution` exists, which returns all the files that will be copied by `archive_and_copy`. `clean` will use this method to only remove files that where copied.

File: mx.py [hunk 1]
```diff
@@ -1406,42 +1406,65 @@ def archive_and_copy(name, arcname):
...
+            for name, arcname in self.get_files_to_archive():
+                archive_and_copy(name, arcname)
 
         self.notify_updated()
 
+    def get_files_to_archive(self, reportOutput=True):
+        filesToArchive = []
+        for d in self.archived_deps():
+            if d.isNativeProject():
+                filesToArchive.extend(self.get_files_for_nativeProject(d, False))
```
Comment (rschatz) [inline mx.py]: Shouldn't this be `get_files_for_nativeProject(d, reportOutput)`?

Comment (blu3r4y) [inline mx.py]: Yes, it should. Thank you for that!

**Comment by blu3r4y** [comment]
The previous build failed, because of a dead link (see #126). I rebased the pull request, should be fine now.

**Comment by rschatz** [comment]
There is still a weird failure if you set the output directory to `.`. I'm attaching a repro case:
[mx-test.tar.gz](https://github.com/graalvm/mx/files/744936/mx-test.tar.gz)

This dummy project builds cleanly on the revision that this branch is based on, but fails on this PR:
```
Archiving TEST_NATIVE... [dependency test.native updated]
Process Process-4:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/home/roland/git/mx/mx.py", line 9868, in executeTask
    task.execute()
  File "/home/roland/git/mx/mx.py", line 708, in execute
    self.build()
  File "/home/roland/git/mx/mx.py", line 1329, in build
    self.subject.make_archive()
  File "/home/roland/git/mx/mx.py", line 1410, in make_archive
    archive_and_copy(name, arcname)
  File "/home/roland/git/mx/mx.py", line 1401, in archive_and_copy
    arc.zf.add(name, arcname=arcname)
  File "/usr/lib/python2.7/tarfile.py", line 2009, in add
    tarinfo = self.gettarinfo(name, arcname)
  File "/usr/lib/python2.7/tarfile.py", line 1881, in gettarinfo
    statres = os.lstat(name)
OSError: [Errno 2] No such file or directory: '/home/roland/test/mx-test/./test-output-file'
Archiving TEST_NATIVE failed
1 build tasks failed
```

**Comment by blu3r4y** [comment]
The issue arises because the clean task gets called during some builds, which cleans files needed by the archive task later on. The clean happens [on this line within mx.py](https://github.com/graalvm/mx/blob/ecc49e995882fc0813f91944ae9f952b847be9e5/mx.py#L706).

I assume the behavior of a "clean-for-build-call" is needed for some projects in some way. Thus, I changed the clean method for native projects, such that it will not clean any files, if the clean method is called with `forBuild = true`.

**Comment by gilles-duboscq** [comment]
> The issue arises because the clean task gets called during some builds, which cleans files needed by the archive task later on. The clean happens on this line within mx.py.

That seems to go against the overall design: a built task should have separate inputs and outputs. It should always be possible to delete those outputs (i.e., clean) before (re-)building. A failure there indicates that output and inputs are somehow mixed.

**Comment by blu3r4y** [comment]
Well, that seems to me like a workaround in `BuildTask` is broken now because of a correct `clean`.

However, to fix this, i just have to take care of the `forBuild` argument in the `clean(forBuild)` call. By the way, this is also done in the `NativeBuildTask` [here](https://github.com/graalvm/mx/blob/ecc49e995882fc0813f91944ae9f952b847be9e5/mx.py#L2851-L2859).

My latest commit 4505fe1866338e3d8c947675fb54403e8d1f981f does exactly that. So it won't break a `clean` within `NativeBuildTask` and one can still perform a sound clean with an explicit `mx clean`.

**Comment by blu3r4y** [comment]
Any updates on that? Can we integrate it with fix 4505fe1866338e3d8c947675fb54403e8d1f981f?

**Comment by gilles-duboscq** [comment]
I still think the required workaround points at a problem (which might be an existing problem, not necessarily a problem in your changes) and i would like to look into it before changing more things around native projects. Unfortunately i didn't have much time to look at it. The reason i'd prefer to be sure before merging this is that once we merge this we have to remain backwards compatible with the current behaviour.

I'll try to make some time to look into it this week.

**Comment by blu3r4y** [comment]
I see your point. If the change is too problematic, don't struggle with other issues arising from it. This feature is not mission critical to me ;)

**Comment by blu3r4y** [comment]
I squashed my commits and made a new PR #129.

**Pull request closed without merging**